torchgeo/datasets/caffe.py

```diff
@@ -0,0 +1,293 @@
+            download: if True, download dataset and store it in the root directory
+            checksum: if True, check the MD5 of the downloaded files (may be slow)
+
+        Raises:
+            AssertionError: if ``split`` argument is invalid
+            DatasetNotFoundError: If dataset is not found and *download* is False.
+        """
+        assert split in self.valid_splits, f'split must be one of {self.valid_splits}'
+
+        self.root = root
+        self.split = split
+        self.transforms = transforms
+        self.download = download
+        self.checksum = checksum
+
+        self._verify()
+
+        self.fpaths = glob.glob(
+            os.path.join(
+                self.root,
+                self.zipfilename.replace('.zip', ''),
+                self.mask_dirs[1],
+                self.split,
+                '*.png',
+            )
+        )
+
+        self.ordinal_map_zones = torch.zeros(
+            max(self.px_class_values_zones.keys()) + 1, dtype=torch.long
+        )
+        for ordinal, px_class in enumerate(self.px_class_values_zones.keys()):
+            self.ordinal_map_zones[px_class] = ordinal
+
+        self.ordinal_map_fronts = torch.zeros(
+            max(self.px_class_values_fronts.keys()) + 1, dtype=torch.long
+        )
+        for ordinal, px_class in enumerate(self.px_class_values_fronts.keys()):
+            self.ordinal_map_fronts[px_class] = ordinal
+
+    def __len__(self) -> int:
+        """Return the number of images in the dataset."""
+        return len(self.fpaths)
+
+    def __getitem__(self, idx: int) -> dict[str, Tensor]:
+        """Return the image and mask at the given index.
+
+        Args:
+            idx: index of the image and mask to return
+
+        Returns:
+            dict: a dict containing the image and mask
+        """
+        zones_filename = os.path.basename(self.fpaths[idx])
+        img_filename = zones_filename.replace('_zones_', '_')
+        front_filename = zones_filename.replace('_zones_', '_front_')
+
+        def read_tensor(path: str) -> Tensor:
+            return torch.from_numpy(np.array(Image.open(path)))
+
+        img = (
+            read_tensor(
+                os.path.join(
+                    self.root, self.data_dir, self.image_dir, self.split, img_filename
+                )
+            )
+            .unsqueeze(0)
+            .float()
+        )
+        front_mask = read_tensor(
+            os.path.join(
+                self.root, self.data_dir, self.mask_dirs[0], self.split, front_filename
+            )
+        ).long()
+        zone_mask = read_tensor(
+            os.path.join(
+                self.root, self.data_dir, self.mask_dirs[1], self.split, zones_filename
+            )
+        ).long()
+
+        zone_mask = self.ordinal_map_zones[zone_mask]
+        front_mask = self.ordinal_map_fronts[front_mask]
+
+        sample = {'image': img, 'mask_front': front_mask, 'mask_zones': zone_mask}
```

Comment:
For our trainers, I was actually hoping to have a single 'mask' key with multiple channels instead of splitting them. Both are compatible with Kornia, but the latter is more likely going to be easy to do in a single generic trainer. This is also how [geodl](https://github.com/maxwell-geospatial/geodl) works (although they require all masks to have a channel dimension, even if there is only a single channel).

Reply:
This is tricky with pytorch and torchmetrics I think because usually the convention is either [batch_size, num_classes, height, width] with one-hot encoding, or [batch_size, height, width] but then class indices. With a single mask you'd try to do both.